lonestar/scientific/cpu/sweeps/sweeps.cpp

```diff
@@ -0,0 +1,825 @@
+  // Rather than do a separate pass over the data to zero out the previous
+  // scattering term for the next iteration, just count the number of
+  // times that the old scattering term is used. Once all the computations
+  // for a given round are done (i.e. when this counter hits zero),
+  // it's safe to move the currently accumulating scattering term value
+  // into the previous accumulated scattering variable and
+  // reset the dependency counters that track when the upstream dependencies
+  // for a given cell/direction have completed.
+  // TODO: Could swapping values separately be faster?
+  // TODO: What about pre-computing the number of incoming edges instead
+  // of re-computing it at each iteration?
+  std::atomic<std::size_t> scatter_use_counter = 0u;
+  // TODO: It wouldn't be hard to just add an iteration counter here too
+  // and then just let the whole thing run fully asynchronously. Try that some
+  // time. Termination detection won't be quite as obvious though if you
+  // want to end when the error gets below a threshold.
+  // For the time being though, this app just runs a set number of iterations.
+};
+
+// No fine-grained locks built into the graph.
+// Use atomics for ALL THE THINGS!
+using graph_t =
+    typename galois::graphs::LC_CSR_Graph<node_t,
+                                          edge_t>::with_no_lockable<true>::type;
+
+// The label for a piece of work in the for_each loop
+// that actually runs the sweeping computation.
+struct work_t {
+  std::size_t node_index;
+  std::size_t direction_index;
+};
+
+// Routine to initialize graph topology and face normals.
+auto generate_grid(graph_t& built_graph, std::size_t nx, std::size_t ny,
+                   std::size_t nz) noexcept {
+
+  galois::graphs::FileGraphWriter temp_graph;
+  // Each node represents a grid cell.
+  // Each edge represents a face.
+  // Ghost nodes are added to represent the exterior
+  // of the domain on the other side of each face.
+  // This is for boundary condition handling.
+  std::size_t num_outer_faces = (nx * ny + ny * nz + nx * nz) * 2;
+  std::size_t num_cells       = nx * ny * nz;
+  std::size_t num_nodes       = num_cells + num_outer_faces;
+  std::size_t num_edges       = 6 * nx * ny * nz + num_outer_faces;
+  temp_graph.setNumNodes(num_nodes);
+  temp_graph.setNumEdges(num_edges);
+  temp_graph.setSizeofEdgeData(
+      galois::LargeArray<graph_t::edge_data_type>::size_of::value);
+  galois::LargeArray<graph_t::edge_data_type> edge_data;
+  edge_data.create(num_edges);
+
+  // Interior cells will have degree 6
+  // since they will have either other cells or
+  // ghost cells on every side.
+  // This condition isn't true in irregular meshes,
+  // but that'd be a separate mesh generation routine.
+  temp_graph.phase1();
+  for (std::size_t i = 0; i < nx; i++) {
+    for (std::size_t j = 0; j < ny; j++) {
+      for (std::size_t k = 0; k < nz; k++) {
+        std::size_t id = i * nx * ny + j * ny + k;
```

Comment:
Should be `i * ny * nz + j * nz + k`.

Reply:
Yep. That's right. I'll update it. Thanks!